### ISSUE TITLE
Failing to install elasticsearch reindexing plugin

### ISSUE THREAD

**Issue by jacobwagner** [rcbops/rpc-openstack#1512]
###### Please provide the output to 'git describe --all' for  all deployments where you have confirmed it occurs.
r13.1.0rc2
##### What type of issue is this? Please tick the appropriate box.
- [x] bug - existing functionality that does not work for you as documented
- [ ] feature - new functionality that you would like to see added

##### What is the expected behavior?
ElasticSearch installs all necessary plugins

##### What is actually happening?
Failing to install reindexing plugin, unknown command being passed

##### What were you attempting to do when you encountered this error (give as much detail as possible)? 
Upgrade from liberty to mitaka

##### How can we reproduce this issue?
see above

##### How can we verify your issue is addressed?
ELK play finishes and installs / configures elk correctly

##### What is your evaluation on the importance of this issue, and why?
important, as rpc cannont finish install of ELK stack

```
TASK: [elasticsearch | Drop log_test1 template] *******************************
changed: [node101_elasticsearch_container-c989f372]

TASK: [elasticsearch | Place legacy liberty template script] ******************
changed: [node101_elasticsearch_container-c989f372 -> node101_elasticsearch_container-c989f372]

TASK: [elasticsearch | Install legacy liberty template] ***********************
changed: [node101_elasticsearch_container-c989f372 -> node101_elasticsearch_container-c989f372]

TASK: [elasticsearch | Remove legacy liberty template script] *****************
changed: [node101_elasticsearch_container-c989f372 -> node101_elasticsearch_container-c989f372]

TASK: [elasticsearch | Install Reindexing Plugin] *****************************
failed: [node101_elasticsearch_container-c989f372] => (item=org.codelibs/elasticsearch-reindexing/1.7.0) => {"changed": true, "cmd": ["./plugin", "-install", "org.codelibs/elasticsearch-reindexing/1.7.0"], "delta": "0:00:00.512909", "end": "2016-10-17 21:36:51.800972", "item": "org.codelibs/elasticsearch-reindexing/1.7.0", "rc": 64, "start": "2016-10-17 21:36:51.288063", "warnings": []}
stdout: ERROR: unknown command [-install]. Use [-h] option to list available commands

FATAL: all hosts have already failed -- aborting

PLAY RECAP ********************************************************************
elasticsearch | Upgrade elasticsearch to stable 1.7.x version ---------- 14.97s
elasticsearch | Wait for elasticsearch to restart ----------------------- 6.34s
Wait for container ssh -------------------------------------------------- 5.30s
pip_install | Install PIP ----------------------------------------------- 2.26s
Create ElasticSearch data directory on host ----------------------------- 2.06s
pip_install | Install pip packages -------------------------------------- 1.83s
pip_install | Check apt last update file -------------------------------- 1.27s
elasticsearch | Start elasticsearch ------------------------------------- 1.17s
elasticsearch | Install pip packages for upgrade ------------------------ 0.90s
elasticsearch | Install Reindexing Plugin ------------------------------- 0.62s
           to retry, use: --limit @/root/setup-logging.retry

node101_elasticsearch_container-c989f372 : ok=28   changed=11   unreachable=0    failed=1
```

Seems it should just be install not -install

```
root@node101-elasticsearch-container-c989f372:/usr/share/elasticsearch/bin# ./plugin -h

NAME

    plugin - Manages plugins

SYNOPSIS

    plugin <command>

DESCRIPTION

    Manage plugins

COMMANDS

    install    Install a plugin

    remove     Remove a plugin

    list       List installed plugins

NOTES

    [*] For usage help on specific commands please type "plugin <command> -h"


root@node101-elasticsearch-container-c989f372:/usr/share/elasticsearch/bin# ./plugin install org.codelibs/elasticsearch-reindexing/1.7.0
-> Installing org.codelibs/elasticsearch-reindexing/1.7.0...
Trying https://download.elastic.co/org.codelibs/elasticsearch-reindexing/elasticsearch-reindexing-1.7.0.zip ...
Trying https://search.maven.org/remotecontent?filepath=org/codelibs/elasticsearch-reindexing/1.7.0/elasticsearch-reindexing-1.7.0.zip ...
Downloading ...DONE
Verifying https://search.maven.org/remotecontent?filepath=org/codelibs/elasticsearch-reindexing/1.7.0/elasticsearch-reindexing-1.7.0.zip checksums if available ...
Downloading .DONE
ERROR: Could not find plugin descriptor 'plugin-descriptor.properties' in plugin zip
```

----------------------------------------------------------------------------------------------------------------------------
###### For each pull request, update the branch line to reference it.
- [x] master
- [x] mitaka-13.1

##### Upgrade impact. Consider the impact on moving between major, minor and patch releases.
- [ ] This change does not impact upgrades
- [x] This change addresses a potential upgrade impact
- [ ] The upgrade impact is being tracked by another issue.

The issue tracking the upgrade impact is:

##### Testing impact
- [ ] An issue/pull-request has been create or exists to track the testing gap, and is referenced below
- [ ] This change includes additional testing
- [x] This change is tested by the existing setup

Additional testing links:

##### Release note
- [ ] This change includes a release note.
- [x] This change does not require a release note.

##### Does your change require documentation?
- [ ] Yes, and I submitted an issue to [Docs issues](https://github.com/rackerlabs/docs-rpc/issues "Docs issues")
- [ ] Yes, and I contributed a documentation patch to [docs-rpc](https://github.com/rackerlabs/docs-rpc "docs-rpc")
- [ ] No documentation required for this change

Additional documentation links (please add links if you have ticked yes): https://github.com/rcbops/u-suk-dev/issues/479

**Comment by alextricity25** [comment]
@jacobwagner: 
This looks like it might be a symptom of running the elasticsearch playbook multiple times. In other words, it's possible that you have an elasticsearch version that is greater than 1.7 at the time this task is run.
``./plugin -install org.codelibs/elasticsearch-reindexing/1.7.0``(with the dash) is actually the correct invocation for elasticsearch version 1.7. On later versions(2.x), it would be ``./plugin install <plugin>``(without the dash, which is what you're seeing). Could you run ``dpkg -l | grep elastic`` just to be sure? If it is what I'm thinking, then you might have no other choice then to purge the package, then re-run the ES upgrade.

**Comment by jacobwagner** [comment]
ii  elasticsearch                    2.4.1                             all          Elasticsearch is a distributed RESTful

It would seem you are correct. This still is an issue. Are we going to ask support to purge packages should they have to run the playbooks twice if the ran into an issue on the upgrade?

**Comment by jacobwagner** [comment]
I am also confused as to how my version of elasticsearch isnt 1.7 if that is what we are expecting to be in place. I did run the plays twice as I forgot to change my lvm driver name.

**Comment by alextricity25** [comment]
@jacobwagner. Version 1.7 is kind of a "transient" version that is used to install the reindexing plugin and reindex ES indices. Once that process is done, the ES version is upgraded to 2.4. Could it be possible that the ES upgrade ran all the way through successfully the first time? That would cause your version to stay at 2.4, and upon re-running, cause the invocation of the ``./plugin`` executable to be incorrect. We might need to write in some logic [here](https://github.com/rcbops/rpc-openstack/blob/master/rpcd/playbooks/roles/elasticsearch/tasks/upgrade_elasticsearch_pre.yml#L30-L60) in these tasks to ensure version 1.7 is installed. Or maybe to skip the re-indexing if a later version is installed. 

I would absolutely agree with you that this is still an issue. We shouldn't ask anyone to purge the ES package if they want to re-run the playbook.

**Comment by jacobwagner** [comment]
@alextricity25 Yes. For testing purposes if a play fails i fix what was wrong and rerun the entire script. It is likely that elasticsearch installed correctly and then the plays failed at a later play. I am attempting to fix it, but purging the packages was not the correct answer as there are bindmounted directories for /var/lib and /var/log and apt cannot remove the package due to this issue. I am doing other steps to get back into a good state atm but not being able to run the elasticsearch play twice due to an indexing task seems dangerous

**Comment by git-harry** [comment]
Given the potential for confusion I'm inclined to suggest that we fix this issue before the release.

To summarise:
- it looks like the task file https://github.com/rcbops/rpc-openstack/blob/master/rpcd/playbooks/roles/elasticsearch/tasks/upgrade_elasticsearch_pre.yml cannot be rerun once ElasticSearch has been upgraded i.e. the elasticsearch.yml cannot be rerun with logging_upgrade set to trun..
- if logging_upgrade != true when the playbook is first run there is no way to perform the migration, I'm not sure if anything would be broken by this or how easy it would be to recover.
- logging_upgrade is also used by the playbook logstash.yml and so setting it to false would potentially impact setup-logging.yml when rerunning the playbooks.

The fix would seem to be to replace the var logging_upgrade with a when condition in the elasticsearch role that checked for the appropriate version of ElasticSearch.

To work around this as it currently is, leaving logging_upgrade == false when rerunning the playbook, on the assumption that the tasks in upgrade_elasticsearch_pre.yml completed successfully the first time round would prevent the failure. However that would probably require the individual logging playbooks be used and not setup-logging.yml.

@jacobwagner would it be fair to say you are in favour of this being addressed in another release candidate?

@d34dh0r53 @mattt416 @alextricity25 @evrardjp I think you were the main ones to work on or test the original commit - what are your opinions on this issue, specifically whether or not we should fix it in another release candidate?

/cc @b3rnard0 @mancdaz @claco

**Comment by claco** [comment]
I agree. I think we should fix this tomorrow. Was just waiting on some mini triage on it.

**Comment by mattt416** [comment]
This was raised early on in the ELK patch testing but was unfortunately lost amongst all the issues we ran into.  I think this should be addresses to avoid any issues w/ live deploys.

**Comment by jacobwagner** [comment]
@git-harry i agree with you. I am just worried about leaving it as is. We will accept another RC as i am having to reset this lab anyway.

**Comment by git-harry** [comment]
Seems like there is agreement, moving to approved and marking as expedited.

**Comment by d34dh0r53** [comment]
I'll get working on a fix for this

**Comment by git-harry** [comment]
@d34dh0r53 how are you getting on with this?

**Comment by tomc603** [comment]
@d34dh0r53 is working on a solution to detect multiple runs, and potentially alert the deployer that manual intervention is necessary. Additionally, an ES version check is necessary to prevent pre-upgrade tasks from running.

**Comment by d34dh0r53** [comment]
*Today:*
* Had a lot of issues with my test environment today.  Setting up a new one now.
* Started work on a fix

*What's left to do:*
* Check installed elasticsearch version
* Check if an index is in progress
* Skip upgrade steps if ES version is up to date
* Exit gracefully if the operator needs to intervene

*Tomorrow:*
* Finish test environment
* Finish preliminary patch

*Blockers:*
* Apt issues that are currently affecting upstream

**Comment by jacobwagner** [comment]
adding this bug seen today as it is another elasticsearch issue

```
TASK: [elasticsearch | Stop elasticsearch] ************************************
ok: [node101_elasticsearch_container-3d247513]

TASK: [elasticsearch | Make sure elasticsearch is stopped] ********************
changed: [node101_elasticsearch_container-3d247513]

TASK: [elasticsearch | Wait for elasticsearch port to close] ******************
ok: [node101_elasticsearch_container-3d247513]

TASK: [elasticsearch | Start elasticsearch] ***********************************
changed: [node101_elasticsearch_container-3d247513]

TASK: [elasticsearch | Wait for elasticsearch to restart] *********************
ok: [node101_elasticsearch_container-3d247513]

TASK: [elasticsearch | Drop log_test1 template] *******************************
failed: [node101_elasticsearch_container-3d247513] => {"changed": true, "cmd": "curl -sk -XDELETE 172.29.236.121:9200/_template/log_test1", "delta": "0:00:00.008777", "end": "2016-10-20 17:24:23.514428", "rc": 7, "start": "2016-10-20 17:24:23.505651", "warnings": ["Consider using get_url module rather than running curl"]}

FATAL: all hosts have already failed -- aborting

PLAY RECAP ********************************************************************
elasticsearch | Add legacy apt repositories ---------------------------- 16.39s
elasticsearch | Wait for elasticsearch to restart ----------------------- 8.35s
elasticsearch | Upgrade elasticsearch to stable 1.7.x version ----------- 6.25s
Wait for container ssh -------------------------------------------------- 5.29s
pip_install | Install PIP ----------------------------------------------- 2.41s
pip_install | Install pip packages -------------------------------------- 1.82s
elasticsearch | Install pip packages for upgrade ------------------------ 0.88s
pip_lock_down | Create pip config directory ----------------------------- 0.50s
pip_install | Get Modern PIP -------------------------------------------- 0.47s
pip_lock_down | Drop pip link file(s) ----------------------------------- 0.43s
           to retry, use: --limit @/root/setup-logging.retry

node101_elasticsearch_container-3d247513 : ok=24   changed=9    unreachable=0    failed=1

root@node84:/opt/rpc-openstack# ansible --version
ansible 1.9.4
  configured module search path = None
```

Seems like the module beings used is wrong and is failing on warning instead of failure.

Havent seen this before so it may just be a transient issue that needs retries until successful

**Comment by d34dh0r53** [comment]
I think that task needs some retries, this seems like a transient error to me.

**Comment by d34dh0r53** [comment]
*Today:*
* Spent most of the day fighting with a galera issue in my test environment.  I finally have it to a place where I can test on both a good install, a broken one and a reindex in progress.

*What's left to do:*
* Check installed elasticsearch version
* Check if an index is in progress
* Skip upgrade steps if ES version is up to date
* Exit gracefully if the operator needs to intervene
* Add retries to the curl portion of the play

*Tomorrow:*
* Finish preliminary patch

*Blockers:*
* N/A
